README.md
tabletop_perception_tools
=========================

Some basic 3D vision tools for detecting objects on tabletops


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package tabletop_perception_tools
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1.0.0 (2015-10-21)
------------------
* made block detector
* now with python interface
* removed incorrect liscense
* more functionality, more parameters
* added some basic blob segmentation
* Contributors: mklingen


scripts/tabletop_interface.py
#!/usr/bin/env python

from tabletop_perception_tools.srv import *
import rospy
import sys
import geometry_msgs.msg

def extract_clusters(service_name="tools_server/extract_clusters", 
                     cloud_topic="/head/kinect2/qhd/points",
                     segment_planes=True, 
                     num_planes=1, 
                     plane_distance=0.01, 
                     segment_depth=True, 
                     min_depth=0.8, 
                     max_depth=1.0, 
                     cluster_tolerance=0.01, 
                     min_cluster_size=100, 
                     max_cluster_size=10000):
    rospy.wait_for_service(service_name);
    try:
        service = rospy.ServiceProxy(service_name, ExtractClusters);
        response = service(cloud_topic, segment_planes, num_planes, plane_distance, segment_depth, min_depth, max_depth, cluster_tolerance, min_cluster_size, max_cluster_size);
        return response.clusters;
    except rospy.ServiceException, e:
        print "Service call failed: %s"%e
        
def find_blocks(service_name="tools_server/find_blocks", 
                     cloud_topic="/head/kinect2/qhd/points",
                     segment_planes=True, 
                     num_planes=1, 
                     plane_distance=0.015, 
                     segment_depth=True, 
                     min_depth=0.9, 
                     max_depth=1.5, 
                     cluster_tolerance=0.005, 
                     min_cluster_size=50, 
                     max_cluster_size=300,
                     segment_box=True,
                     box_min = [-0.5, 0.1, 0.6],
                     box_max = [0.5, 0.8, 1.5]):
    print "waiting for service..."
    rospy.wait_for_service(service_name);
    print "Calling service..."
    try:
        box_min_pt = geometry_msgs.msg.Point();
        box_min_pt.x = box_min[0];
        box_min_pt.y = box_min[1];
        box_min_pt.z = box_min[2];
        box_max_pt = geometry_msgs.msg.Point();
        box_max_pt.x = box_max[0];
        box_max_pt.y = box_max[1];
        box_max_pt.z = box_max[2];
        
        service = rospy.ServiceProxy(service_name, FindBlocks);
        response = service(cloud_topic, segment_planes, num_planes, plane_distance, segment_depth, min_depth, max_depth, cluster_tolerance, min_cluster_size, max_cluster_size, segment_box, box_min_pt, box_max_pt);
        return response.blocks;
    except rospy.ServiceException, e:
        print "Service call failed: %s"%e


if __name__ == "__main__":
    print find_blocks();


include/tabletop_perception_tools/Geometry.h
#ifndef GEOMETRY_H_
#define GEOMETRY_H_



namespace pcl_helpers
{
    typedef Eigen::Vector2f Vec2;
    typedef Eigen::Vector3f Vec3;
    typedef Eigen::Vector4f Vec4;
    typedef Eigen::Matrix3f Mat3x3;
    typedef Eigen::Matrix4f Mat4x4;
    typedef Eigen::Affine3f Transform;
    typedef Eigen::Quaternionf Quaternion;

    typedef std::vector<Vec2, Eigen::aligned_allocator<Vec2> > Vec2List;
    typedef std::vector<Vec3, Eigen::aligned_allocator<Vec3> > Vec3List;
    typedef std::vector<Vec4, Eigen::aligned_allocator<Vec4> > Vec4List;
    typedef std::vector<Mat3x3, Eigen::aligned_allocator<Mat3x3> > Mat3x3List;
    typedef std::vector<Mat4x4, Eigen::aligned_allocator<Mat4x4> > Mat4List;
    typedef std::vector<Transform, Eigen::aligned_allocator<Transform> > TransformList;
    typedef std::vector<Quaternion, Eigen::aligned_allocator<Quaternion> > QuaternionList;


    struct OBB
    {
            EIGEN_MAKE_ALIGNED_OPERATOR_NEW;
            Vec3 min;
            Vec3 max;
            Transform transform;
    };
}



#endif // GEOMETRY_H_ 


include/tabletop_perception_tools/PCLHelpers.h
#ifndef PCLHELPERS_H_
#define PCLHELPERS_H_

#include <ros/ros.h>
#include <pcl/ModelCoefficients.h>
#include <pcl/point_types.h>
#include <pcl/io/pcd_io.h>
#include <pcl/filters/extract_indices.h>
#include <pcl/filters/voxel_grid.h>
#include <pcl/filters/passthrough.h>
#include <pcl/features/normal_3d.h>
#include <pcl/kdtree/kdtree.h>
#include <pcl/sample_consensus/method_types.h>
#include <pcl/sample_consensus/model_types.h>
#include <pcl/segmentation/sac_segmentation.h>
#include <pcl/segmentation/extract_clusters.h>
#include <pcl/common/transforms.h>
#include <pcl/common/common.h>
#include <pcl/visualization/pcl_visualizer.h>

#include "Geometry.h"

namespace pcl_helpers
{
    template<typename PointT> struct Cluster
    {
            typename pcl::PointCloud<PointT>::Ptr points;
            OBB bounds;
    };

    template<typename PointT> typename pcl::PointCloud<PointT>::Ptr FilterBoundingBox(const typename pcl::PointCloud<PointT>::Ptr& points, const Vec3& min, const Vec3& max)
    {
        typename pcl::PointCloud<PointT>::Ptr cloud_filtered(new pcl::PointCloud<PointT>(*points));

        //calculate bounding box
        float xmin = min.x();
        float xmax = max.x();
        float ymin = min.y();
        float ymax = max.y();
        float zmin = min.z();
        float zmax = max.z();

        // Create the filtering object x
        pcl::PassThrough<PointT> pass_x;
        pass_x.setInputCloud (points);
        pass_x.setFilterFieldName ("x");
        pass_x.setFilterLimits (xmin, xmax);
        pass_x.filter (*cloud_filtered);

        // Create the filtering object y
        pcl::PassThrough<PointT> pass_y;
        pass_y.setInputCloud (cloud_filtered);
        pass_y.setFilterFieldName ("y");
        pass_y.setFilterLimits (ymin, ymax);
        pass_y.filter (*cloud_filtered);

        // Create the filtering object z
        pcl::PassThrough<PointT> pass_z;
        pass_z.setInputCloud (cloud_filtered);
        pass_z.setFilterFieldName ("z");
        pass_z.setFilterLimits (zmin, zmax);
        pass_z.filter (*cloud_filtered);

        return cloud_filtered;
    }


    template<typename PointT> OBB ComputeBounds(const typename pcl::PointCloud<PointT>::Ptr& points)
    {
        // compute principal direction
        Vec4 centroid;
        pcl::compute3DCentroid(*points, centroid);
        Mat3x3 covariance;
        computeCovarianceMatrixNormalized(*points, centroid, covariance);

        Eigen::SelfAdjointEigenSolver<Mat3x3> eigen_solver(covariance, Eigen::ComputeEigenvectors);
        Mat3x3 eigDx = eigen_solver.eigenvectors();
        eigDx.col(2) = eigDx.col(0).cross(eigDx.col(1));

        // move the points to the that reference frame
        Mat4x4 p2w(Mat4x4::Identity());
        p2w.block<3,3>(0,0) = eigDx.transpose();
        p2w.block<3,1>(0,3) = -1.f * (p2w.block<3,3>(0,0) * centroid.head<3>());
        pcl::PointCloud<PointT> cPoints;
        pcl::transformPointCloud(*points, cPoints, p2w);

        PointT min_pt, max_pt;
        pcl::getMinMax3D(cPoints, min_pt, max_pt);
        const Vec3 mean_diag = 0.5f*(max_pt.getVector3fMap() + min_pt.getVector3fMap());

        // final transform
        const Quaternion qfinal(eigDx);
        const Vec3 tfinal = eigDx*mean_diag + centroid.head<3>();


        OBB toReturn;
        toReturn.transform.linear() = qfinal.toRotationMatrix();
        toReturn.transform.translation() = centroid.head<3>();
        toReturn.min = min_pt.getVector3fMap();
        toReturn.max = max_pt.getVector3fMap();

        return toReturn;

    }

    template <typename PointT> bool DetectPlane(const typename pcl::PointCloud<PointT>::Ptr& points, std::vector<float>& plane_params,
             float distanceThreshold = 0.015f, int methodType = pcl::SAC_RANSAC, int modelType = pcl::SACMODEL_PLANE)
    {
        pcl::SACSegmentation<PointT> seg;
        pcl::PointIndices::Ptr inliers(new pcl::PointIndices);
        pcl::ModelCoefficients::Ptr coefficients(new pcl::ModelCoefficients);
        typename pcl::PointCloud<PointT>::Ptr cloud_plane(new pcl::PointCloud<PointT>());
        seg.setOptimizeCoefficients(true);
        seg.setModelType(modelType);
        seg.setMethodType(methodType);
        seg.setMaxIterations(100);
        seg.setDistanceThreshold(distanceThreshold);

        typename pcl::PointCloud<PointT>::Ptr cloud_filtered(new pcl::PointCloud<PointT>(*points));
        std::vector<int> nanIndex;
        pcl::removeNaNFromPointCloud(*cloud_filtered,*cloud_filtered, nanIndex);
        size_t nr_points = cloud_filtered->points.size();

        //Get largest plane component params
        inliers->indices.clear();
        seg.setInputCloud(cloud_filtered);
        seg.segment(*inliers, *coefficients);

        if (inliers->indices.size() == 0)
        {
            //No inliers -> plane not detected
            return false;
        }
        else
        {
            plane_params = coefficients->values;
            return true;
        }


    }   


    template <typename PointT> typename pcl::PointCloud<PointT>::Ptr RemoveLargePlanes(const typename pcl::PointCloud<PointT>::Ptr& points,
            int numPlanes = 1, float distanceThreshold = 0.015f, int methodType = pcl::SAC_RANSAC, int modelType = pcl::SACMODEL_PLANE)
    {

        // Create the segmentation object for the planar model and set all the parameters
        pcl::SACSegmentation<PointT> seg;
        pcl::PointIndices::Ptr inliers(new pcl::PointIndices);
        pcl::ModelCoefficients::Ptr coefficients(new pcl::ModelCoefficients);
        typename pcl::PointCloud<PointT>::Ptr cloud_plane(new pcl::PointCloud<PointT>());
        seg.setOptimizeCoefficients(true);
        seg.setModelType(modelType);
        seg.setMethodType(methodType);
        seg.setMaxIterations(100);
        seg.setDistanceThreshold(distanceThreshold);


        typename pcl::PointCloud<PointT>::Ptr cloud_filtered(new pcl::PointCloud<PointT>(*points));
        std::vector<int> nanIndex;
        pcl::removeNaNFromPointCloud(*cloud_filtered,*cloud_filtered, nanIndex);
        size_t nr_points = cloud_filtered->points.size();
        for (int k = 0; k < numPlanes; k++)
        {
            inliers->indices.clear();
            // Segment the largest planar component from the remaining cloud
            seg.setInputCloud(cloud_filtered);
            seg.segment(*inliers, *coefficients);
            if (inliers->indices.size() == 0)
            {
                return cloud_filtered;
            }

            // Extract the planar inliers from the input cloud
            pcl::ExtractIndices<PointT> extract;
            extract.setInputCloud(cloud_filtered);
            extract.setIndices(inliers);

            // Remove the planar inliers, extract the rest
            extract.setNegative(true);
            extract.filter(*cloud_filtered);
        }

        pcl::removeNaNFromPointCloud(*cloud_filtered,*cloud_filtered, nanIndex);

        return cloud_filtered;

    }

    float randf()
    {
        return (float)(rand()) / (float)(RAND_MAX);
    }

    template <typename PointT> void ExtractClusters(const typename pcl::PointCloud<PointT>::Ptr& points, std::vector<Cluster<PointT> >* clusters, float tolerance = 0.2f, int minClusterSize = 100, int maxClusterSize = 25000)
    {
        // Creating the KdTree object for the search method of the extraction
        typename pcl::search::KdTree<PointT>::Ptr tree(new pcl::search::KdTree<PointT>);
        tree->setInputCloud(points);

        std::vector<pcl::PointIndices> cluster_indices;
        pcl::EuclideanClusterExtraction<PointT> ec;
        ec.setClusterTolerance(tolerance);
        ec.setMinClusterSize(minClusterSize);
        ec.setMaxClusterSize(maxClusterSize);
        ec.setSearchMethod(tree);
        ec.setInputCloud(points);
        ec.extract(cluster_indices);

        /*
        pcl::visualization::PCLVisualizer* viewer (new pcl::visualization::PCLVisualizer ("3D Viewer"));
        viewer->setBackgroundColor (0, 0, 0);
        viewer->addCoordinateSystem (1.0);
        viewer->initCameraParameters ();
        viewer->addPointCloud(points, "FullPoints");
        */
        //int j = 0;
        for (std::vector<pcl::PointIndices>::const_iterator it = cluster_indices.begin(); it != cluster_indices.end(); ++it)
        {
            typename pcl::PointCloud<PointT>::Ptr cloud_cluster(new pcl::PointCloud<PointT>);
            for (std::vector<int>::const_iterator pit = it->indices.begin(); pit != it->indices.end(); pit++)
            {
                cloud_cluster->points.push_back(points->points[*pit]);
            }
            cloud_cluster->width = cloud_cluster->points.size();
            cloud_cluster->height = 1;
            cloud_cluster->is_dense = true;

            Cluster<PointT> cluster;
            cluster.points = cloud_cluster;
            cluster.bounds = ComputeBounds<PointT>(cloud_cluster);
            clusters->push_back(cluster);
            /*
            std::stringstream ss;
            ss << j;

            pcl::visualization::PointCloudColorHandlerCustom<PointT> single_color (points, randf() * 255, randf() * 255, randf() * 255);
            viewer->addPointCloud<PointT> (cloud_cluster, single_color, "cluster " + ss.str());
            j++;
            */
        }

        /*
        while (!viewer->wasStopped ())
        {
          viewer->spinOnce (100);
          boost::this_thread::sleep (boost::posix_time::microseconds (100000));
        }
        delete viewer;
        */

    }
}


#endif // PCLHELPERS_H_ 


src/tools_server.cpp
#include <ros/ros.h>
#include <pcl/ModelCoefficients.h>
#include <pcl/point_types.h>
#include <pcl/io/pcd_io.h>
#include <pcl/filters/extract_indices.h>
#include <pcl/filters/voxel_grid.h>
#include <pcl/features/normal_3d.h>
#include <pcl/kdtree/kdtree.h>
#include <pcl/sample_consensus/method_types.h>
#include <pcl/sample_consensus/model_types.h>
#include <pcl/segmentation/sac_segmentation.h>
#include <pcl/segmentation/extract_clusters.h>

#include <pcl_ros/point_cloud.h>
#include <pcl/conversions.h>

#include <sensor_msgs/PointCloud2.h>
#include <tabletop_perception_tools/PCLHelpers.h>
#include <tabletop_perception_tools/ExtractClusters.h>
#include <tabletop_perception_tools/FindBlocks.h>
#include <tabletop_perception_tools/DetectPlane.h>
#include <visualization_msgs/MarkerArray.h>

using namespace tabletop_perception_tools;

ros::Publisher vis_publisher;
ros::Publisher cloud_pub;
ros::NodeHandle* nodeHandle = 0x0;
typedef pcl::PointXYZRGB Point;
typedef pcl::PointCloud<Point> PointCloud;
PointCloud::ConstPtr lastPointCloud;
bool pointCloudDirty = true;

void PointCloudCallback(PointCloud::ConstPtr pointCloud)
{
    lastPointCloud = pointCloud;
    pointCloudDirty = false;
}

bool ExtractClustersService(ExtractClusters::Request& request, ExtractClusters::Response& response)
{
    std::string topic = request.point_cloud_topic;
    pointCloudDirty = true;
    ros::Subscriber sub = nodeHandle->subscribe(topic, 1, &PointCloudCallback);
    ros::Duration timeout(15);
    ros::Time begin = ros::Time::now();
    ROS_INFO("Waiting for point cloud %s", topic.c_str());
    while (pointCloudDirty)
    {
        usleep(10000);
        ROS_INFO("Waiting...");
        if (ros::Time::now() - begin > timeout)
        {
            ROS_WARN("Timeout exceeded waiting for point cloud %s", topic.c_str());
            return false;
        }
    }

    ROS_INFO("Got point cloud. Extracting");
    PointCloud::Ptr pointCloud(new PointCloud(*lastPointCloud));

    if (request.segment_depth)
    {
        ROS_INFO("Segmenting depth");
        pointCloud = pcl_helpers::FilterBoundingBox<Point>(pointCloud, pcl_helpers::Vec3(-1000, -1000, request.min_depth), pcl_helpers::Vec3(1000, 1000, request.max_depth));
    }

    if (request.segment_planes)
    {
        ROS_INFO("Removing a plane");
        pointCloud = pcl_helpers::RemoveLargePlanes<Point>(pointCloud, request.num_planes, request.plane_distance);

    }

    ROS_INFO("Removing the clusters");
    std::vector<pcl_helpers::Cluster<Point> > clusters;
    pcl_helpers::ExtractClusters<Point>(pointCloud, &clusters, request.cluster_tolerance, request.min_cluster_size, request.max_cluster_size);

    ROS_INFO("Extracted %lu clusters", clusters.size());
    visualization_msgs::MarkerArray markers;

    for (size_t i = 0; i < clusters.size(); i++)
    {
        pcl_helpers::Cluster<Point>& cluster = clusters.at(i);

        tabletop_perception_tools::Cluster cluster_msg;
        cluster_msg.header.stamp = ros::Time::now(); // TODO: get the right timestamp
        cluster_msg.header.frame_id = pointCloud->header.frame_id;
        cluster_msg.max.x = cluster.bounds.max.x();
        cluster_msg.max.y = cluster.bounds.max.y();
        cluster_msg.max.z = cluster.bounds.max.z();

        cluster_msg.min.x = cluster.bounds.min.x();
        cluster_msg.min.y = cluster.bounds.min.y();
        cluster_msg.min.z = cluster.bounds.min.z();

        cluster_msg.pose.position.x = cluster.bounds.transform.translation().x();
        cluster_msg.pose.position.y = cluster.bounds.transform.translation().y();
        cluster_msg.pose.position.z = cluster.bounds.transform.translation().z();

        pcl_helpers::Quaternion quat(cluster.bounds.transform.linear());

        cluster_msg.pose.orientation.x = quat.x();
        cluster_msg.pose.orientation.y = quat.y();
        cluster_msg.pose.orientation.z = quat.z();
        cluster_msg.pose.orientation.w = quat.w();

        visualization_msgs::Marker marker;
        marker.color.r = 1;
        marker.color.g = 0.8;
        marker.color.b = 0;
        marker.color.a = 0.5;
        marker.header.frame_id = cluster_msg.header.frame_id;
        marker.type = visualization_msgs::Marker::CUBE;
        marker.pose = cluster_msg.pose;
        marker.scale.x = (cluster_msg.max.x - cluster_msg.min.x);
        marker.scale.y = (cluster_msg.max.y - cluster_msg.min.y);
        marker.scale.z = (cluster_msg.max.z - cluster_msg.min.z);
        marker.header.stamp = ros::Time::now();
        markers.markers.push_back(marker);

        pcl::toROSMsg(*cluster.points, cluster_msg.points);
        response.clusters.push_back(cluster_msg);
    }

    vis_publisher.publish(markers);

    ROS_INFO("Shutting down...");
    sub.shutdown();

    ROS_INFO("Done.");
    return true;
}

bool FindBlocksService(FindBlocks::Request& request, FindBlocks::Response& response)
{
    std::string topic = request.point_cloud_topic;
    pointCloudDirty = true;
    ros::Subscriber sub = nodeHandle->subscribe(topic, 1, &PointCloudCallback);
    ros::Duration timeout(20);
    ros::Time begin = ros::Time::now();
    ROS_INFO("Waiting for point cloud %s", topic.c_str());
    while (pointCloudDirty)
    {
        usleep(10000);
        ROS_INFO("Waiting...");
        if (ros::Time::now() - begin > timeout)
        {
            ROS_WARN("Timeout exceeded waiting for point cloud %s", topic.c_str());
            return false;
        }
    }

    ROS_INFO("Got point cloud. Extracting");
    PointCloud::Ptr pointCloud(new PointCloud(*lastPointCloud));

    if (request.segment_depth)
    {
        ROS_INFO("Segmenting depth");
        pointCloud = pcl_helpers::FilterBoundingBox<Point>(pointCloud, pcl_helpers::Vec3(-1000, -1000, request.min_depth), pcl_helpers::Vec3(1000, 1000, request.max_depth));
        ROS_INFO("Filtered cloud has %lu points\n", pointCloud->size());
    }

    if (request.segment_box)
    {
        ROS_INFO("Segmenting box");
        pointCloud = pcl_helpers::FilterBoundingBox<Point>(pointCloud, pcl_helpers::Vec3(request.box_min.x, request.box_min.y, request.box_min.z), pcl_helpers::Vec3(request.box_max.x, request.box_max.y, request.box_max.z));
        ROS_INFO("Filtered cloud has %lu points\n", pointCloud->size());
    }

    if (request.segment_planes)
    {
        ROS_INFO("Removing planes");
        pointCloud = pcl_helpers::RemoveLargePlanes<Point>(pointCloud, request.num_planes, request.plane_distance);
        ROS_INFO("Filtered cloud has %lu points\n", pointCloud->size());
    }

    ROS_INFO("Filtered cloud has %lu points\n", pointCloud->size());
    cloud_pub.publish(pointCloud);

    ROS_INFO("Removing the clusters");
    std::vector<pcl_helpers::Cluster<Point> > clusters;
    pcl_helpers::ExtractClusters<Point>(pointCloud, &clusters, request.cluster_tolerance, request.min_cluster_size, request.max_cluster_size);

    ROS_INFO("Extracted %lu clusters", clusters.size());
    visualization_msgs::MarkerArray markers;
    for (size_t i = 0; i < clusters.size(); i++)
    {
        pcl_helpers::Cluster<Point>& cluster = clusters.at(i);

        tabletop_perception_tools::Block block_msg;
        block_msg.header.stamp = ros::Time::now(); // TODO: get the right timestamp
        block_msg.header.frame_id = pointCloud->header.frame_id;


        block_msg.pose.position.x = cluster.bounds.transform.translation().x();
        block_msg.pose.position.y = cluster.bounds.transform.translation().y();
        block_msg.pose.position.z = cluster.bounds.transform.translation().z();

        pcl_helpers::Quaternion quat(cluster.bounds.transform.linear());

        block_msg.pose.orientation.x = quat.x();
        block_msg.pose.orientation.y = quat.y();
        block_msg.pose.orientation.z = quat.z();
        block_msg.pose.orientation.w = quat.w();

        float avg_r = 0.0f;
        float avg_g = 0.0f;
        float avg_b = 0.0f;
        float avg_a = 1.0f;
        for (size_t j = 0; j < cluster.points->size(); j++)
        {
            const Point& p = cluster.points->at(j);
            uint32_t rgb = *((int*)(&p.rgb));
            uint8_t r = (rgb >> 16) & 0x0000ff;
            uint8_t g = (rgb >> 8)  & 0x0000ff;
            uint8_t b = (rgb)       & 0x0000ff;
            avg_r += r;
            avg_g += g;
            avg_b += b;
        }

        avg_r /= cluster.points->size();
        avg_g /= cluster.points->size();
        avg_b /= cluster.points->size();

        avg_r /= 255.0f;
        avg_g /= 255.0f;
        avg_b /= 255.0f;

        block_msg.avg_color.r = avg_r;
        block_msg.avg_color.g = avg_g;
        block_msg.avg_color.b = avg_b;
        block_msg.avg_color.a = avg_a;

        visualization_msgs::Marker marker;
        marker.color.r = avg_r;
        marker.color.g = avg_g;
        marker.color.b = avg_b;
        marker.color.a = avg_a;
        marker.header.frame_id = block_msg.header.frame_id;
        marker.type = visualization_msgs::Marker::CUBE;
        marker.pose = block_msg.pose;
        marker.scale.x = 0.025f;
        marker.scale.y = 0.025f;
        marker.scale.z = 0.025f;
        marker.header.stamp = ros::Time::now();
        marker.id = i;
        markers.markers.push_back(marker);

        response.blocks.push_back(block_msg);
    }

    vis_publisher.publish(markers);

    ROS_INFO("Shutting down...");
    sub.shutdown();

    ROS_INFO("Done.");
    return true;
}


bool DetectPlaneService(DetectPlane::Request& request, DetectPlane::Response& response)
{
    std::string topic = request.point_cloud_topic;
    pointCloudDirty = true;
    ros::Subscriber sub = nodeHandle->subscribe(topic, 1, &PointCloudCallback);
    ros::Duration timeout(15);
    ros::Time begin = ros::Time::now();
    ROS_INFO("Waiting for point cloud %s", topic.c_str());
    while (pointCloudDirty)
    {
        usleep(10000);
        ROS_INFO("Waiting...");
        if (ros::Time::now() - begin > timeout)
        {
            ROS_WARN("Timeout exceeded waiting for point cloud %s", topic.c_str());
            return false;
        }
    }

    ROS_INFO("Got point cloud. Extracting");
    PointCloud::Ptr pointCloud(new PointCloud(*lastPointCloud));
    std::vector<float> plane_params;

    //Try to detect plane
    bool plane_detected = pcl_helpers::DetectPlane<Point>(pointCloud,plane_params);

    if(plane_detected)
    {
        response.a = plane_params[0];
        response.b = plane_params[1];
        response.c = plane_params[2];
        response.d = plane_params[3];
        response.ok = true;

        ROS_INFO_STREAM("Found plane! Coefficients: ["<<response.a<<" "<<response.b<<" "<<response.c<<" "<<response.d<<"]");
    }
    else
    {
        response.ok = false;
        ROS_INFO("Did not find plane!");
    }

    return true;
}



int main(int argc, char** argv)
{
    ros::init(argc, argv, "tools_server");
    ros::NodeHandle node("~");
    nodeHandle = &node;
    ros::ServiceServer server = node.advertiseService(std::string("extract_clusters"), &ExtractClustersService);
    ros::ServiceServer block_server = node.advertiseService(std::string("find_blocks"), &FindBlocksService);
    ros::ServiceServer plane_server = node.advertiseService(std::string("detect_planes"), &DetectPlaneService);
    vis_publisher = node.advertise<visualization_msgs::MarkerArray>(std::string("cluster_markers"), 1);
    cloud_pub = node.advertise<PointCloud>(std::string("filtered_cloud"), 1);
    ros::AsyncSpinner spinner(4);
    spinner.start();
    ros::waitForShutdown();
    return 0;
}
